adriengivry/Overload
# Add CMake suppport.

Opening a new PR for CMake. All I did was copy the CMake files from the Linux+CMake branch to this one. This branch builds the generator (CMake files) on Windows and Linux.

- **What your PR is bringing to Overload (Description of the changes, use cases)**
This PR adds CMake support by creating CMakeLists.txt files across the projects. This makes it easier to port Overload to other platforms.

- **Requirements (CMake version)**
Minimum CMake version is 3.13.

- **Limitations (Supported configurations)**
Overload cannot be run immediately after build - it must be installed first and than launched.

- **Quick guide to test your changes**
To test this branch in Windows using Visual Studio 2019 you will have to
1. Download and install CMake from their [website](https://cmake.org/download/) or using the Visual Studio Installer (under Individual components)
2. Open it inside Visual Studio by 'Open a local folder' inside "Overload/Sources/Overload"
3. Wait for it to configure (see created out folder)
4. Build All.
5. Run the editor by clicking the arrow next to the run button and selecting 'OvEditor.exe (Install)' and running.
![image](https://user-images.githubusercontent.com/38976370/85341640-f4c22d00-b4e8-11ea-9068-9b7adfb0177b.png)
6. Everything should hopefully work.

- **How to keep CMake support up-to-date with Overload future updates**
The current structure of the CMake files for each project is structured like this (OvWindowing used for example):
OvWindowing/CMakeLists.txt - 'Main' project file - it links dependencies with current project and adds two subdirectories containing more CMake files.

OvWindowing/include/CMakeLists.txt - Adds header and inline files to current project. If you create a new file you add it to the list here.

OvWindowing/src/CMakeLists.txt - Same as above but with source files.

The root CMake list found inside "Sources/Overload/CMakeLists.txt" sets some global variables,  includes CMake scripts to find external dependencies (see cmake folder inside the root directory), adds all sub-projects using ```add_subdirectory(OvWindowing)``` and finally installs all the projects.

Knowing all that..
- creating a new class is as simple as listing it inside include and src CMake lists.
- to create a new project you add the subdirectory inside the root CMakeLists.txt and to install TARGETS (it should be pretty obvious from what is already done)
- to add a new external library you first create a new cmake script file to "find" the library, include it inside the root CMakeLists.txt and add it to target_link_link_libraries for the project that requires it.
